File: examples/python/agg-order/handlers/create_order.py
"""Handler for CreateOrder command."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import types_pb2 as types
from errors import CommandRejectedError
from proto import domains_pb2 as domains

from .state import OrderState


def handle_create_order(
    command_book: types.CommandBook,
    command_any: Any,
    state: OrderState,
    seq: int,
) -> types.EventBook:
    """Handle CreateOrder command."""
    if state.exists():
        raise CommandRejectedError("Order already exists")

    cmd = domains.CreateOrder()
    command_any.Unpack(cmd)

    if not cmd.customer_id:
        raise CommandRejectedError("Customer ID is required")
    if not cmd.items:
        raise CommandRejectedError("Order must have at least one item")

    subtotal = sum(item.quantity * item.unit_price_cents for item in cmd.items)

    event = domains.OrderCreated(
        customer_id=cmd.customer_id,
        subtotal_cents=subtotal,
        created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )
    event.items.extend(cmd.items)

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return types.EventBook(
        cover=command_book.cover,
        pages=[
            types.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )

File: examples/rust/agg-order/src/handlers/create_order.rs
//! Handler for CreateOrder command.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{CreateOrder, OrderCreated, OrderState};
use common::{
    decode_command, now, require_not_empty, require_not_exists, require_positive, ProtoTypeName,
    Result,
};

use crate::errmsg;
use crate::state::build_event_response;

/// Handle the CreateOrder command.
pub fn handle_create_order(
    command_book: &CommandBook,
    command_data: &[u8],
    state: &OrderState,
    next_seq: u32,
) -> Result<EventBook> {
    require_not_exists(&state.customer_id, errmsg::ORDER_EXISTS)?;

    let cmd: CreateOrder = decode_command(command_data)?;

    require_not_empty(&cmd.items, errmsg::ITEMS_REQUIRED)?;
    for item in &cmd.items {
        require_positive(item.quantity, errmsg::QUANTITY_POSITIVE)?;
    }

    let subtotal: i32 = cmd
        .items
        .iter()
        .map(|i| i.quantity * i.unit_price_cents)
        .sum();

    let event = OrderCreated {
        customer_id: cmd.customer_id.clone(),
        items: cmd.items.clone(),
        subtotal_cents: subtotal,
        created_at: Some(now()),
        customer_root: cmd.customer_root.clone(),
        cart_root: cmd.cart_root.clone(),
    };

    Ok(build_event_response(
        state,
        command_book.cover.clone(),
        next_seq,
        &OrderCreated::type_url(),
        event,
    ))
}
